### ISSUE TITLE
eslint-configured globals should not be reported as unused

### ISSUE THREAD

**Issue by broofa** [eslint/eslint#7818]
* **ESLint Version:** v3.11.1
* **Node Version:** v6.9.1
* **npm Version:** 4.0.5

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

See `cat .eslintrc.json`, below

**What did you do? Please include the actual source code causing the issue.**

```
kieffer@Roberts-MacBook-Pro$ cat .eslintrc.json
{
  "rules": {
    "no-unused-vars": "warn",
    "no-undef": "error"
  }
}

kieffer@Roberts-MacBook-Pro$ cat foo.js
/* global foo */

kieffer@Roberts-MacBook-Pro$ eslint foo.js

/Users/kieffer/work/foo.js
  1:11  warning  'foo' is defined but never used  no-unused-vars

✖ 1 problem (0 errors, 1 warning)
```

**What did you expect to happen?**

Since `foo` is never actually declared in real code, it should not be reported as being unused.

**What actually happened? Please include the actual, raw output from ESLint.**

The comment-configured global `foo` is reported as unused, despite not being defined anywhere in real code.

**"Just remove the unused var(s) from the configuration comment, you moron!"**

The use-case for this is that I have code that dynamically creates helper methods for each of our DB models. Something along the lines of ...
```js
// Dynamically create some helper functions based on DB model classes
for (var modelName in models) global['is' + modelName] = function(thing) {...}

// Tell eslint about the functions we just created
/* global isUser, isAccount, isSession, etc... */
```

This configuration comment avoids the more serious `no-undef` errors if/when a helper function is used.  But I end up with `no-unused-vars` warnings for the ones that aren't.  It's a lose-lose situation, as far as getting eslint to shut up goes. :-/

Make sense?

**Comment by nzakas** [comment]
This is working as designed. The inline comment defining the global is treated the same as var. If you're not using a global that you have defined in the file, then there's no reason to have it defined there. 

Also, the global comment is hoisted, so it's applied before your code is evaluated. I'm afraid this is just an improper use of the global comment, and I'd suggest just removing the "global" part of that comment to make it inert to ESLint.

**Comment by broofa** [comment]
@nzakas  I don't think hoisting behavior needs to be a consideration in this discussion.  Let's just assume any real references to the to globals in question appear after the configuration comment.

FWIW, if `foo` is declared in .eslintrc.json via `"globals": {"foo": true}` then there's no error.  That some globals trigger the `no-unused-var` rule but others don't is arguably inconsistent.

> I'd suggest just removing the "global" part of that comment to make it inert to ESLint.

I don't know what this means.  Making the configuration comment inert just means I get `no-undef` errors instead of `no-unused-var` errors. 😕   I guess I'm reading this as you saying, "Just delete the comment, you moron!" which I tried to address in my original report.  Am I missing something?  Is there some flavor of configuration comment that lets me declare vars w/in a file w/out triggering `no-unused-var` warnings if they aren't referenced anywhere?

**Comment by not-an-aardvark** [comment]
@broofa I think @nzakas is saying to only declare the globals that actually get used in the file.

So instead of doing something like this:

```js
/* global foo, bar, baz */

console.log(foo);
console.log(baz);
```

You should do this:


```js
/* global foo, baz */

console.log(foo);
console.log(baz);
```

**Comment by broofa** [comment]
@platinumazure In my use case, I'm inside a node.js module file, where setting properties of the `global` object allows you to reference them without accessing them off `global`.

So in my case I'm hanging these properties off `global` so that I can write code further down that looks like `isUser(...)` rather than having to do the more literal and cumbersome `global.isUser(...)`.  So from the perspective of eslint when validating this file, they're effectively global.

**Comment by broofa** [comment]
@not-an-aardvark Okay... so effectively the "Just remove the unused var(s) from the configuration comment, you moron".  😄 

*shrug*  I don't know how to respond to that other than to say it's really weird that a file with no actual variable declarations can generate unused variable warnings.

**Comment by not-an-aardvark** [comment]
Sorry, maybe I'm not understanding the issue correctly. Why is it necessary to declare a global in a file if it's not used? I see that your example in the original issue involves dynamically assigning properties to the `global` object, but I don't see how that would prevent you from only declaring globals that actually get used.

**Comment by broofa** [comment]
@not-an-aardvark I want to create globals based on the structure of what I know is available.  It's the same rational as you might have for declaring `window` and `document` globals.  These are properties I know to be present in the global space.  Whether or not they're actually referenced is immaterial.

**Comment by not-an-aardvark** [comment]
If the globals are available for all files rather than a particular file, it might be better to specify them in your config:

```yml
globals:
  window: true
  document: true
  isUser: true
  # ...
```

**Comment by broofa** [comment]
@not-an-aardvark I'm hanging these off the [CommonJS global object](https://nodejs.org/api/globals.html#globals_global) in a specific module file, so they're only available in the context of one particular file. Hence, why I'm reluctant to do this in my config file.

**Comment by not-an-aardvark** [comment]
The Node global object is shared among all modules. If you add properties to it in one module, they will also be accessible in all other modules.

**Comment by broofa** [comment]
@not-an-aardvark TIL  And feel like a bit of an idiot.  Thanks for correcting me.  'Not sure why I thought `global` was scoped to a module.

I need to ponder the implications of that for my own code.  It may be I should be doing this another way.

I do think there's still an issue here.  The eslint docs imply inline `global` and config file `globals` should be interchangeable:

> You can define global variables either using comments inside of a file or in the configuration file.

 ([link](http://eslint.org/docs/user-guide/configuring#specifying-globals)).  That they're not is (to me) a little weird.  But given my own misunderstanding of what my code is doing I'll concede this isn't worth attending to for the time being.  Sorry to have troubled you folks.
